packages/next/src/client/components/app-router.tsx

```diff
@@ -592,8 +603,46 @@ export default function AppRouter({
   globalErrorComponent: ErrorComponent
   assetPrefix: string
 }) {
+  if (process.env.__NEXT_APP_NAV_FAIL_HANDLING) {
+    // this if is only for DCE of the feature flag not conditional
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useEffect(() => {
+      const uncaughtExceptionHandler = (
+        evt: ErrorEvent | PromiseRejectionEvent
+      ) => {
+        const error = 'reason' in evt ? evt.reason : evt.error
+        // if we have an unhandled exception/rejection during
+        // a navigation we fall back to a hard navigation to
+        // attempt recovering to a good state
+        if (
+          window.next.__pendingUrl &&
+          createHrefFromUrl(new URL(window.location.href)) !==
+            window.next.__pendingUrl
```

Comment:
`__pendingUrl` is a full URL and `createHrefFromURL` (poorly named) just returns `window.location.pathname + window.location.search`, so this would always fallback to hard navigation right?

Reply:
Good catch, updated to run both through the util for consistency